server/src-lib/Hasura/GraphQL/Execute.hs

```diff
@@ -117,100 +106,101 @@ instance MonadGQLExecutionCheck m => MonadGQLExecutionCheck (Tracing.TraceT m) w
-      throw500 "unexpected custom type for top level field"
+  -> m (C.GQLContext, QueryParts)
+getExecPlanPartial userInfo sc queryType req =
+  (getGCtx ,) <$> getQueryParts req
   where
     roleName = _uiRole userInfo
 
+    contextMap =
+      case queryType of
+        ET.QueryHasura -> scGQLContext sc
+        ET.QueryRelay  -> scRelayContext sc
+
+    defaultContext =
+      case queryType of
+        ET.QueryHasura -> scUnauthenticatedGQLContext sc
+        ET.QueryRelay  -> scUnauthenticatedRelayContext sc
+
+    getGCtx :: C.GQLContext
+    getGCtx =
+      case Map.lookup roleName contextMap of
+        Nothing  -> defaultContext
+        Just (C.RoleContext frontend backend) ->
+          case _uiBackendOnlyFieldAccess userInfo of
+            BOFAAllowed    -> fromMaybe frontend backend
+            BOFADisallowed -> frontend
+
+    -- | Depending on the request parameters, fetch the correct typed operation
+    -- definition from the GraphQL query
+    getQueryParts
+      :: MonadError QErr m
+      => GQLReqParsed
+      -> m QueryParts
+    getQueryParts (GQLReq opNameM q _varValsM) = do
+      let (selSets, opDefs, _fragDefsL) = G.partitionExDefs $ unGQLExecDoc q
+      case (opNameM, selSets, opDefs) of
+        (Just opName, [], _) -> do
+          let n = _unOperationName opName
+              opDefM = find (\opDef -> G._todName opDef == Just n) opDefs
+          onNothing opDefM $ throw400 ValidationFailed $
+            "no such operation found in the document: " <> showName n
+        (Just _, _, _)  ->
+          throw400 ValidationFailed $ "operationName cannot be used when " <>
+          "an anonymous operation exists in the document"
+        (Nothing, [selSet], []) ->
+          return $ G.TypedOperationDefinition G.OperationTypeQuery Nothing [] [] selSet
+        (Nothing, [], [opDef])  ->
+          return opDef
+        (Nothing, _, _) ->
+          throw400 ValidationFailed $ "exactly one operation has to be present " <>
+          "in the document when operationName is not specified"
+
+-- The graphql query is resolved into a sequence of execution operations
+data ResolvedExecutionPlan m
+  = QueryExecutionPlan (EPr.ExecutionPlan (m EncJSON, EQ.GeneratedSqlMap) EPr.RemoteCall (G.Name, J.Value))
+  -- ^ query execution; remote schemas and introspection possible
+  | MutationExecutionPlan (EPr.ExecutionPlan (m EncJSON, HTTP.ResponseHeaders) EPr.RemoteCall (G.Name, J.Value))
+  -- ^ mutation execution; only __typename introspection supported
+  | SubscriptionExecutionPlan EL.LiveQueryPlan
+  -- ^ live query execution; remote schemas and introspection not supported
+
+validateSubscriptionRootField
+  :: MonadError QErr m
+  => C.QueryRootField v -> m (C.SubscriptionRootField v)
+validateSubscriptionRootField = \case
+  C.RFDB x -> pure $ C.RFDB x
+  C.RFAction (C.AQAsync s) -> pure $ C.RFAction s
+  C.RFAction (C.AQQuery _) -> throw400 NotSupported "action queries are not supported over subscriptions"
```

Comment:
```suggestion
  C.RFAction (C.AQQuery _) -> throw400 NotSupported "query actions cannot be run as a subscription"
```

Reply:
This is the message that's being thrown in current master

Reply:
Done.